### PR TITLE
Safely propagate the state of InvocationContextImpl accross threads

### LINKED ISSUE #4355: Safely propagate the state of InvocationContextImpl accross threads
The object does not have to be thread-safe though. Use case - SmallRye FT has an interceptor that calls `InvocationContext#process()` on a different thread.

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #4355

### CODE REVIEW

**Review by kenfinnigan — APPROVED** [review]

**Comment by mkouba** [comment]
There are some weird issues I need to investigate... :-(

**Comment by mkouba** [comment]
It was actually a PEBKAC error... time to finish today ;-).

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InvocationContextImpl.java [hunk 1]
```diff
@@ -82,9 +87,9 @@ public static InvocationContextImpl aroundConstruct(Constructor<?> constructor,
 
     private final Constructor<?> constructor;
 
-    private Object[] args;
+    private final List<Object> args;
 
-    private int position;
+    private final AtomicInteger position;
```
Comment (stuartwdouglas) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InvocationContextImpl.java]: Can you change this to a volatile integer with an AtomicIntegerUpdater instead? This object can be created a lot so it is better to save the allocation.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InvocationContextImpl.java [hunk 2]
```diff
@@ -82,9 +87,9 @@ public static InvocationContextImpl aroundConstruct(Constructor<?> constructor,
 
     private final Constructor<?> constructor;
 
-    private Object[] args;
+    private final List<Object> args;
```
Comment (stuartwdouglas) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InvocationContextImpl.java]: Why not just use a volatile field for the array?

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InvocationContextImpl.java (hunk 1 above)
Comment (stuartwdouglas) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InvocationContextImpl.java]: TBH I don't think you even need the updater, if you have multiple threads calling next() concurrently then you are into undefined territory anyway. Just making the field volatile should be enough to make sure the state is visible to a new thread.

**Comment by mkouba** [comment]
Will be superseded by a PR for https://github.com/quarkusio/quarkus/issues/4552.

**Pull request closed without merging**